File: ScaleformUI_Csharp/Menus/Pause Menus/Elements/Items/MissionItem.cs
using ScaleformUI.Elements;
using ScaleformUI.LobbyMenu;
using ScaleformUI.Menu;
using ScaleformUI.PauseMenu;
using ScaleformUI.PauseMenus.Elements.Columns;
using ScaleformUI.Scaleforms;
using System.Drawing;

namespace ScaleformUI.PauseMenus.Elements.Items
{
    public delegate void MissionActivated(PlayerListTab tab, MissionsListColumn column, MissionItem item);
    public class MissionItem : PauseMenuItem
    {
        private bool enabled = true;
        internal int type = 0;
        internal KeyValuePair<string, string> customLeftBadge;
        internal KeyValuePair<string, string> customRightBadge;
        internal bool rIcChecked;
        public bool Jumpable { get; internal set; }
        public new string Label { get => base.Label.Label; set => base.Label = value; }
        public MissionsListColumn ParentColumn { get; internal set; }
        public SColor MainColor { get; private set; } = SColor.FromHudColor(HudColor.HUD_COLOUR_PAUSE_BG);
        public SColor HighlightColor { get; private set; } = SColor.FromHudColor(HudColor.HUD_COLOUR_WHITE);
        public BadgeIcon LeftIcon { get; private set; } = BadgeIcon.NONE;
        public SColor LeftIconColor { get; private set; } = SColor.FromHudColor(HudColor.HUD_COLOUR_WHITE);
        public BadgeIcon RightIcon { get; private set; } = BadgeIcon.NONE;
        public SColor RightIconColor { get; private set; } = SColor.FromHudColor(HudColor.HUD_COLOUR_WHITE);
        public bool RightIconChecked { get; private set; }
        public bool Hovered { get; internal set; }
        public bool Enabled
        {
            get => enabled; set
            {
                enabled = value;
            }
        }

        public event MissionActivated MissionActivated;
        public void ActivateMission(PlayerListTab tab)
        {
            MissionActivated?.Invoke(tab, ParentColumn, this);
        }

        public MissionItem(string label) : this(label, SColor.FromHudColor(HudColor.HUD_COLOUR_PAUSE_BG), SColor.FromHudColor(HudColor.HUD_COLOUR_WHITE)) { }

        public MissionItem(string label, SColor mainColor, SColor highlightColor) : base(label)
        {
            MainColor = mainColor;
            HighlightColor = highlightColor;
            type = 0;
            customLeftBadge = new KeyValuePair<string, string>("", "");
            customRightBadge = new KeyValuePair<string, string>("", "");
        }

        public virtual void SetLeftIcon(BadgeIcon icon, SColor color)
        {
            LeftIcon = icon;
            LeftIconColor = color;
        }
        public virtual void SetRightIcon(BadgeIcon icon, SColor color, bool @checked = false)
        {
            RightIcon = icon;
            RightIconColor = color;
            RightIconChecked = @checked;
        }

        public virtual void SetCustomLeftIcon(string txd, string txn)
        {
            LeftIcon = BadgeIcon.CUSTOM;
            customLeftBadge = new KeyValuePair<string, string>(txd, txn);
        }

        public virtual void SetCustomRightIcon(string txd, string txn, bool @checked = false)
        {
            RightIcon = BadgeIcon.CUSTOM;
            customRightBadge = new KeyValuePair<string, string>(txd, txn);
            rIcChecked = @checked;
        }
    }
    public class MissionSeparatorItem : MissionItem
    {
        /// <summary>
        /// Use it to create an Empty item to separate Mission Items
        /// </summary>
        public MissionSeparatorItem(string title, bool jumpable) : base(title)
        {
            type = 1;
            Jumpable = jumpable;
            customLeftBadge = new KeyValuePair<string, string>("","");
            customRightBadge = new KeyValuePair<string, string>("", ""); 
        }

        public override void SetLeftIcon(BadgeIcon badge, SColor color)
        {
            throw new Exception("MissionSeparatorItem cannot have a left badge.");
        }
        public override void SetRightIcon(BadgeIcon badge, SColor color, bool @checked = false)
        {
            throw new Exception("MissionSeparatorItem cannot have a right badge.");
        }
    }
}

File: ScaleformUI_Lua/src/Menus/PauseMenus/Elements/Items/MissionItem.lua
MissionItem = {}
MissionItem.__index = MissionItem
setmetatable(MissionItem, { __index = PauseMenuItem })
MissionItem.__call = function() return "MissionItem" end

---@class MissionItem
---@field Label string
---@field ParentColumn MissionListColumn
---@field enabled boolean
---@field MainColor SColor
---@field HighlightColor SColor
---@field LeftIcon BadgeStyle
---@field LeftIconColor SColor
---@field RightIcon BadgeStyle
---@field RightIconColor SColor
---@field RightIconChecked boolean
---@field _Selected boolean
---@field hovered boolean
---@field Activated fun(item:MissionItem)
---@field Hovered fun(val:boolean)
---@field Enabled fun(bool:boolean)
---@field SetLeftIcon fun(icon:BadgeStyle, color:SColor)
---@field SetRightIcon fun(icon:BadgeStyle, color:SColor, checked:boolean)


---@param label string
---@param mainColor SColor | nil
---@param highlightColor SColor | nil
---@return table
function MissionItem.New(label, mainColor, highlightColor)
    local base = PauseMenuItem.New(label)
    base.type = 0
    base.enabled = true
    base.MainColor = mainColor or SColor.HUD_Pause_bg
    base.HighlightColor = highlightColor or SColor.HUD_White
    base.LeftIcon = BadgeStyle.NONE
    base.LeftIconColor = SColor.HUD_White
    base.RightIcon = BadgeStyle.NONE
    base.RightIconColor = SColor.HUD_White
    base.RightIconChecked = false
    base._Selected = false
    base.hovered = false
    base.customLeftIcon = {TXD="",TXN=""}
    base.customRightIcon = {TXD="",TXN=""}
    base.Activated = function(item)
    end
    return setmetatable(base, MissionItem)
end

function MissionItem:Hovered(val)
    if val == nil then
        return self.hovered
    else
        self.hovered = val
    end
end

function MissionItem:Enabled(bool)
    if bool == nil then
        return self.enabled
    else
        self.enabled = bool
        if self.ParentColumn ~= nil and self.ParentColumn:visible() then
            self.ParentColumn:UpdateSlot(IndexOf(self.ParentColumn.Items, self))
        end
    end
end

function MissionItem:SetLeftIcon(icon, color)
    self.LeftIcon = icon
    self.LeftIconColor = color
    if self.ParentColumn ~= nil and self.ParentColumn:visible() then
        self.ParentColumn:UpdateSlot(IndexOf(self.ParentColumn.Items, self))
    end
end

function MissionItem:SetRightIcon(icon, color, checked)
    self.RightIcon = icon
    self.RightIconColor = color
    self.RightIconChecked = checked or false
    if self.ParentColumn ~= nil and self.ParentColumn:visible() then
        self.ParentColumn:UpdateSlot(IndexOf(self.ParentColumn.Items, self))
    end
end

function MissionItem:SetCustomLeftIcon(txd, txn)
    self.LeftIcon = BadgeStyle.CUSTOM
    self.customLeftIcon = {TXD=txd, TXN=txn}
    if self.ParentColumn ~= nil and self.ParentColumn:visible() then
        self.ParentColumn:UpdateSlot(IndexOf(self.ParentColumn.Items, self))
    end
end
function MissionItem:SetCustomRightIcon(txd, txn, checked)
    self.RightIcon = BadgeStyle.CUSTOM
    self.RightIconChecked = checked or false
    self.customRightIcon = {TXD=txd, TXN=txn}
    if self.ParentColumn ~= nil and self.ParentColumn:visible() then
        self.ParentColumn:UpdateSlot(IndexOf(self.ParentColumn.Items, self))
    end
end

function MissionItem:Selected(bool)
    if bool == nil then
        return self._Selected
    else
        self._Selected = bool
    end
end
